### ISSUE TITLE
Wrong ParamRate for 

### ISSUE THREAD

**Issue by khodzha** [servo/media#338]
[The spec](https://webaudio.github.io/web-audio-api/#AudioParam-attributes) says that `playbackRate` and `detune` of BufferSourceNode must be k-rate.
But right now [they](https://github.com/servo/media/blob/master/audio/buffer_source_node.rs#L104) [are](https://github.com/servo/media/blob/master/audio/buffer_source_node.rs#L109) a-rate.

Also spec says must be impossible to change them from k-rate to a-rate.

I suggest to put `ParamType` into `Param` struct so its possible to enforce these constraints directly in `Param`

**Comment by jdm** [comment]
Using types to avoid this problem sounds like a wise solution.

**Comment by Manishearth** [comment]
Types can't fix this, the automationrate is mutable. It's some specific parameters where it's not.

(In general: most knobs in WebAudio are tweakable in general and have constraints in specific cases)

**Comment by Manishearth** [comment]
We actually sample these as k-rate anyway.

**Comment by khodzha** [comment]
I tried doing something like this:
https://github.com/khodzha/media/commit/108d7c6335e0de1c18f1cfafb971a22ef6a7e152#diff-c073ca9541b87e1d6a2a9d87c644a81cR146

but idk how to go from `AudioNodeEngine` to `PanningNode` to query `panning_model`

**Comment by Manishearth** [comment]
Right, those checks belong in servo's script crate, at best we can have some debug assertions here IMO

**Comment by khodzha** [comment]
> We actually sample these as k-rate anyway

you mean now ARate is implemented as KRate?

**Comment by Manishearth** [comment]
@khodzha hm? I mean that we sample them at the beginning of the block, not once per tick/frame.

**Comment by khodzha** [comment]
https://github.com/servo/media/pull/340

I've exposed node types via audio context, then it can be used in SetAutomationRate to do these checks
